### DISCUSSION TITLE
RISC-V

### DISCUSSION THREAD

**Discussion by cederom** [blackmagic-debug/blackmagic#1751] [Q&A] [answered]
Hello world :-) Does BlackMagic support RISC-V debug? :-)

**Comment by dragonmux** [comment]
Yes, current `main` has fully working support for the GD32VF103, branch `feature/esp32-c3-support` mostly working support for the ESP32-C3, and there are other branches contributors have made for supporting WCH targets that use their own low-level protocol.

The in-tree support supports talking to any JTAG-connected 32-bit RISC-V device and has provisional support for 64-bit ones. This includes support for scan, attach, breakpoints, watchpoints and other major debugging requirements. Support for Flashing parts requires target-specific support being written but all the major parts necessary for that should already exist.

`main` represents what will be BMD v2.0

**Reply by cederom** [reply]
Thanks @dragonmux :-)

So generic support for 32-bit and 64-bit RV is almost here? Assuming 0.13 or upcoming 1.0.0 rv debug spec [1]?

Would it be possible to (generic) debug bigger SoC like U74 based JH7110 [2] on VisionFive2 [3] or STAR64 [4] SBC out-of-the-box with v2.3 hardware?

What is the architecture for target / RV debug specifications? Is it some sort of module / library?

What is the HAL architecture of BlackMagic? Would it be possible and okay for you to add BlackMagic code as part of (Apache-2.0 licensed) NuttX RTOS [5] application / submodule then add features to the upstream or you prefer to keep the firmware running only on your hardware?

[1] https://github.com/riscv/riscv-debug-spec/releases
[2] https://doc-en.rvspace.org/Doc_Center/jh7110.html
[3] https://doc-en.rvspace.org/Doc_Center/visionfive_2.html
[4] https://wiki.pine64.org/wiki/STAR64
[5] https://nuttx.apache.org

**Reply by dragonmux** [reply, marked as answer]
The support is indeed for the v0.13 spec w/ some small amounts of support for the v1.0 spec though obviously we can't actually support that yet as it's not ratified. v0.11 is a whole headache that is on the roadmap to deal with but not yet.

The listed SoCs should all do JTAG, so BMP v2.3 hardware + either a custom firmware build that enables the RISC-V support, or any BMDA build based on latest `main` will work and should provide full generic RISC-V 32-bit debug.

It's neither a module nor a library - rather, it's done with layers, like the target layer, architecture layer, etc - [see the guide on adding a new target](https://black-magic.org/hacking/target-drivers.html) for some more information on what target support looks like - note the environment does not care about the underlying architecture as that's the point of the architecture layer, which is also split into layers - [riscv32](https://github.com/blackmagic-debug/blackmagic/blob/659db420f703c10ddd50bb4bcec327c23792a0e0/src/target/riscv32.c), [riscv_debug](https://github.com/blackmagic-debug/blackmagic/blob/659db420f703c10ddd50bb4bcec327c23792a0e0/src/target/riscv_debug.c), and [riscv_jtag_dtm](https://github.com/blackmagic-debug/blackmagic/blob/659db420f703c10ddd50bb4bcec327c23792a0e0/src/target/riscv_jtag_dtm.c). `riscv_debug.c` there handles the debug spec side of things and how exactly to talk with a generic RISC-V target regardless of bitness.

Black Magic Debug's firmware is built on top of libopencm3. The entire thing is GPLv3+ licensed (individual bits are BSD-3-Clause and MIT, but the overall license winds up GPLv3+). NB it is not designed nor built for an RTOS environment, but other projects like [farpatch](https://github.com/farpatch/farpatch) which are built on it do use it in such an environment. However, it's designed in a way where you plug in your ~wierd and~ exotic platform into the platforms API and the project doesn't know or care the difference - as long as the API contracts are fulfilled, you can do whatever you have to in a platform. This makes it very pluggable.

**Reply by cederom** [reply]
Thank you @dragonmux for your time and all of the information, amazing project! :-)
